content/things/webgpu.md

```diff
@@ -0,0 +1,586 @@
+
+With the advent of ML, neural networks, and dare I say cryptocurrencies, GPUs have shown that they can be useful for more than just drawing triangles on the screen. Using GPUs for calculations of any kind is often called General-Purpose GPU or GPGPU, and WebGL 1 is not great at that. If you wanted to process arbitrary data to the GPU you have to encode it as a texture, decode it in a shader, do your calculations and then re-encode the result as a texture. WebGL 2 made this a lot easier with [Transform Feedback], but WebGL2 wasn’t supported in Safari until September 2021, so it wasn’t really an option (while most other browers supported WebGL2 since January 2017).
+
+## WebGPU
+
+Outside of the web, a new generation of graphics APIs have established themselves which expose a more low-level interface to graphics cards and accommodate multiple requirements. On the one hand, GPUs are almost ubiquitous now. Even our mobile devices have capable GPUs built in. As a result, _both_ modern graphics programming (like ray tracing) and GPGPU use-cases are increasingly common. On the other hand, most of our devices have multi-core processors, so being able to interact with the GPU from multiple threads can be an important optimization vector. While they were at it, they also revisited some design decisions of the previous APIs and front-loaded a lot of the validation work that GPUs have to do, allowing the developer to squeeze more performance out of the GPUs.
+
+The most popular of the next-gen GPU APIs are Vulcan by the Khronos Group, Metal by Apple and DirectX 12 by Microsoft. To bring these new capabilities to the web, WebGPU was born. It abstracts the idiosyncrasies of these different APIs and has been standardized in the W3C with all major browser vendors having a seat at the table. Due to their low-level nature and their sheer power, WebGPU has a bit of a learning curve and is relatively heavy on the setup, but I’ll try to break it down as best I can. As the title of this blog post implies, I will restrict myself to GPGPU use-case without using WebGPU to draw pixels. I think it may make it easier to grasp as an introduction to WebGPU.
+
+### Adapters and Devices
+
+WebGPU’s first abstractions that you come into contact with are _adapters_ and (logical) _devices_.
+
+<figure>
+  <img loading="lazy" width=418 height=393 src="./architecture.svg">
+  <figcaption>Layers of abstraction, from physical GPUs to logical devices.</figcaption>
+</figure>
+
+A _physical_ device is the GPU itself, often distinguished between built-in GPUs and discrete GPUs. Commonly, any given device has exactly one GPU, but it also possible to have two or more GPUs. For example, Microsoft’s SurfaceBook famously has a low-powered integrated GPU and a high-performance discrete GPU between which the operating system will switch on demand.
```

Comment:
Not sure where it would be best to put it, if you want to say anything at all, but you could also mention that in some cases we can also expose software adapters (called fallback adapters in the WebGPU spec) to ensure the widest compatibility possible.